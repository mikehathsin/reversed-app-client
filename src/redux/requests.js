// Ducks: Redux Reducer Bundles
// https://github.com/erikras/ducks-modular-redux
// A proposal for bundling reducers, action types and actions when using Redux

import axios from "axios";
import { v4 as uuidv4 } from "uuid";

const { REACT_APP_API_URL = "http://localhost:8000/iecho" } = process.env;

// Actions
const SEND_TEXT_REQUEST = "SEND_TEXT_REQUEST";
const SEND_TEXT_SUCCESS = "SEND_TEXT_SUCCESS";
const SEND_TEXT_FAILURE = "SEND_TEXT_FAILURE";

// Action creators
const sendTextRequest = () => {
  return {
    type: SEND_TEXT_REQUEST,
  };
};
const sendTextSuccess = (payload) => {
  return {
    type: SEND_TEXT_SUCCESS,
    payload,
  };
};
const sendTextFailure = (payload) => {
  return {
    type: SEND_TEXT_FAILURE,
    payload,
  };
};

// Thunk
export const sendText = (text) => {
  return (dispatch) => {
    dispatch(sendTextRequest());

    axios
      .get(`${REACT_APP_API_URL}?text=${text}`)
      .then((res) => {
        dispatch(
          sendTextSuccess({
            id: uuidv4(),
            datetime: new Date().toLocaleTimeString(),
            ...res.data,
          })
        );
      })
      .catch((err) => {
        dispatch(sendTextFailure(err.response?.data.error || err.message));
      });
  };
};

const initialState = {
  isLoading: false,
  error: "",
  data: [],
};

// Reducer
export const requests = (state = initialState, action) => {
  switch (action.type) {
    case SEND_TEXT_REQUEST:
      return {
        ...state,
        isLoading: true,
        error: "",
      };

    case SEND_TEXT_SUCCESS:
      return {
        ...state,
        isLoading: false,
        data: [action.payload, ...state.data],
      };

    case SEND_TEXT_FAILURE:
      return {
        ...state,
        isLoading: false,
        error: action.payload,
      };

    default:
      return state;
  }
};
