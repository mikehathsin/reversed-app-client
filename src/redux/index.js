import { applyMiddleware, combineReducers, createStore } from "redux";
import thunk from "redux-thunk";

import { requests } from "./requests";

const rootReducer = combineReducers({
  requests,
});

export const store = createStore(rootReducer, applyMiddleware(thunk));
