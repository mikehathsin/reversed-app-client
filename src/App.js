import Col from "react-bootstrap/Col";
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";

import { useDispatch, useSelector } from "react-redux";

import { Form } from "./components/form";
import { Header } from "./components/header";
import { Spinner } from "./components/spinner";
import { TextList } from "./components/text-list";
import { sendText } from "./redux/requests";

export const App = () => {
  const dispatch = useDispatch();
  const { isLoading, error, data } = useSelector(
    (reducers) => reducers.requests
  );

  const handleSubmit = (text, setText) => {
    return (event) => {
      event.preventDefault();
      dispatch(sendText(text));
      setText("");
    };
  };

  return (
    <Container fluid>
      <Row className="py-4 bg-danger justify-content-center">
        <Col md={6}>
          <Form onSubmit={handleSubmit} isLoading={isLoading} />
        </Col>
      </Row>

      <Row className="py-4 justify-content-center">
        <Col md={6}>
          <Header error={error} />
        </Col>
      </Row>

      <Row className="mb-4 justify-content-center">
        <Col md={6}>{isLoading ? <Spinner /> : <TextList data={data} />}</Col>
      </Row>
    </Container>
  );
};
