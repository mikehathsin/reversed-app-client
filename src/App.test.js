import { render } from "@testing-library/react";
import { Provider } from "react-redux";

import { App } from "./App";
import { store } from "./redux";

describe("App tests", () => {
  it("renders without crashing", () => {
    const component = render(
      <Provider store={store}>
        <App />
      </Provider>
    );
    component.getByText("Results:");
    expect(component.container).toMatchSnapshot();
  });
});
