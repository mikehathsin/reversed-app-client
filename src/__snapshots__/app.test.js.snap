// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests renders without crashing 1`] = `
<div>
  <div
    class="container-fluid"
  >
    <div
      class="py-4 bg-danger justify-content-center row"
    >
      <div
        class="col-md-6"
      >
        <form
          class="d-flex"
        >
          <input
            class="form-control"
            placeholder="Insert text"
            value=""
          />
          <button
            class="ms-2 btn btn-primary"
            type="submit"
          >
            Send
          </button>
        </form>
      </div>
    </div>
    <div
      class="py-4 justify-content-center row"
    >
      <div
        class="col-md-6"
      >
        
        <h1>
          Results:
        </h1>
      </div>
    </div>
    <div
      class="mb-4 justify-content-center row"
    >
      <div
        class="col-md-6"
      >
        <div
          class="vstack gap-2"
        >
          <h4>
            No data. Send your first text!
          </h4>
        </div>
      </div>
    </div>
  </div>
</div>
`;
